### PR TITLE
Refactor of UDF test

### PR DESCRIPTION
**Pull request by southpolesteve**
Refactor of just the UDF test. Made a new PR to discuss some of the changes instead of committing to the existing WIP branch. I'll make some inline notes.

### CODE REVIEW

File: src/client/UserDefinedFunction/UserDefinedFunctions.ts [hunk 1]
```diff
@@ -10,31 +10,51 @@ export class UserDefinedFunctions {
     private client: CosmosClient;
     constructor(public readonly container: Container) {
         this.client = this.container.database.client;
-     }
+    }
```
Comment (southpolesteve) [inline src/client/UserDefinedFunction/UserDefinedFunctions.ts]: All changes in this files are cosmetic.

File: src/test/functional/udf.spec.ts [hunk 2]
```diff
@@ -1,154 +1,167 @@
 import * as assert from "assert";
-import * as Stream from "stream";
-import {
```
Comment (southpolesteve) [inline src/test/functional/udf.spec.ts]: Looks like the lint step doesn't check for unused imports? I think we should add that in a different PR.

File: src/test/functional/udf.spec.ts [hunk 3]
```diff
@@ -1,154 +1,167 @@
...
+
+    beforeEach(async function() {
         this.timeout(10000);
-        try {
-            await TestHelpers.removeAllDatabases(host, masterKey);
-        } catch (err) {
-            throw err;
-        }
+        // remove all databases from the endpoint before each test
+        await TestHelpers.removeAllDatabases(client);
```
Comment (southpolesteve) [inline src/test/functional/udf.spec.ts]: Per DM discussion, I think we can omit the try-catch blocks here and mocha will handle errors properly. I'll verify.

File: src/test/functional/udf.spec.ts [hunk 4]
```diff
@@ -1,154 +1,167 @@
...
+                );
             }
         });
 
-        it("nativeApi Should do UDF CRUD operations successfully rid based with upsert", async function () {
-            try {
-                await udfCRUDTest(false, true);
-            } catch (err) {
-                throw err;
-            }
```
Comment (southpolesteve) [inline src/test/functional/udf.spec.ts]: Same comment as https://github.com/Azure/azure-cosmos-js/pull/19/files#r197522725

File: src/test/functional/udf.spec.ts [hunk 5]
```diff
@@ -1,154 +1,167 @@
...
+                id: dbId
+            });
+
+            // create container
+            await client.databases
+                .getDatabase(dbId)
+                .containers.create({ id: "sample container" });
+
+            // read udfs
+            const container = await client.databases
```
Comment (southpolesteve) [inline src/test/functional/udf.spec.ts]: By keeping the `container` reference around, seems like `TestHelper` becomes unnecessary. At least for this test.

File: src/test/functional/udf.spec.ts [hunk 6]
```diff
@@ -1,154 +1,167 @@
...
+            const {
+                result: udfs
+            } = await container.userDefinedFunctions.read().toArray();
+
+            // Steve: Is this assert unecessary given types?
+            assert.equal(Array.isArray(udfs), true);
+
+            // create a udf
+            const beforeCreateUdfsCount = udfs.length;
+            const udfDefinition: any = {
```
Comment (southpolesteve) [inline src/test/functional/udf.spec.ts]: Typescript doesn't seem to like this and how it gets used later. I don't quite understand why yet, but it is the last error preventing me from compiling the file.

File: src/client/UserDefinedFunction/UserDefinedFunctions.ts [hunk 7]
```diff
@@ -10,31 +10,51 @@ export class UserDefinedFunctions {
...
 
     public getUserDefinedFunction(id: string): UserDefinedFunction {
         return new UserDefinedFunction(this.container, id);
     }
 
-    public query(query: SqlQuerySpec, options?: FeedOptions): QueryIterator<UserDefinedFunctionDefinition> {
-        return this.client.documentClient.queryUserDefinedFunctions(this.container.url, query, options);
+    public query(
+        query: SqlQuerySpec,
+        options?: FeedOptions
```
Comment (christopheranderson) [inline src/client/UserDefinedFunction/UserDefinedFunctions.ts]: this violates the current tslint config

Comment (christopheranderson) [reply src/client/UserDefinedFunction/UserDefinedFunctions.ts]: try running tslint on just this file and check it out

File: src/test/functional/udf.spec.ts (hunk 2 above)
Comment (christopheranderson) [reply src/test/functional/udf.spec.ts]: Agreed. tslint doesn't check for that with the current settings. it should

File: src/test/functional/udf.spec.ts [hunk 8]
```diff
@@ -1,153 +1,234 @@
...
+                result: res
+            } = await container.userDefinedFunctions
+                .getUserDefinedFunction(replacedUdf.id)
+                .delete();
+
+            // read udfs after deletion
             try {
-                await udfCRUDTest(true, true);
+                const {
+                    result: badudf
```
Comment (christopheranderson) [inline src/test/functional/udf.spec.ts]: Is this prettier's formatting? Seems odd to me.

**Comment by southpolesteve** [comment]
👍 I think whatever default prettier setup I have is not working with
tslint. Will figure out how to fix that

**Comment by southpolesteve** [comment]
Fixed linting+formatting issues and cherry-picked onto https://github.com/Azure/azure-cosmos-js/pull/16

**Pull request closed without merging**